### PR TITLE
[9.0] [ADD] Self service scale for loose goods and products sold by weight

### PR DESCRIPTION
**Pull request by Fkawala**
This add-on allows you to run a grocery in a Bring Your Own Container (BYOC) schema. With this add-on, your customers will be able to buy loose goods using their own reusable containers. This add-on also enables customers to weight their vegetables and fruits before they checkout. To weight fruit and vegetable before checkout will speed up the checkout and make your customer happier.

The **self-service** BYOC scheme has six steps:
1. The customer weighs her-his pot and sticks the tare barcode onto the pot.
2. The customer puts loose goods into the labeled pot.
3. The customer scans the tare barcode. POS saves the tare value.
4. The customer weighs the pot with loose goods inside. POS computes the price to pay from gross weight and tare weight.
5. The customer gets a price barcode label indicating the price to pay.
6. During checkout, the cashier swiftly scans the price barcode label.

This add-on requires you to have a barcode label printer and an electronic scale. Your customers are expected to use this add-on in a self-service way. This add-on adds multiple news screens that are displayed only on the selected POS.

This add-on displays only the products that are meant to be weighted with a scale.

This addon depends on: 
- point_of_sale
- pos_barcode_tare

You can test with the following tare barcode: 0700000001235 it'll tare for 0.123kg.
Demo data include an article with a weight code bar: 2112345000008.

Below are screens for a POS with this add-on enabled.

The default screen:
default_screen

The product screen:
product_screen

The tare barcode scanning screen:
tare_scan_screen

A price barcode looks like:
barcode

The configuration screen of POS has one new option:
config

### CODE REVIEW

**Comment by Fkawala** [comment]
Hello @legalsylvain this is related to #436! Could you have a look whenever you have the time? Thank you in advance + have a nice day.

**Comment by Fkawala** [comment]
Hello @robinkeunen, if you have some spare time to review this PR that would be great! Thank you in advance!

**Comment by Fkawala** [comment]
FYI. this addon is conflicting with pos_empty_home. I'm not sure which is the best way to solve this conflict. One solution could be to update pos_empty_home for it to have an on/off toggle option in the POS model. I could code that if this is a good enough solution.

**Comment by Fkawala** [comment]
Hello @legalsylvain, I hope you're doing as well as possible given the current circumstances. Do you think that you'll have some spare time to spend on this? All the best and stay safe.

**Comment by legalsylvain** [comment]
Huge work to review. I'll take a time this weel if possible.
I'm in favor to accept this PR, once reviewed, as it is on some "obosolete" v9 branch and to make some work and refactoring on 12 branch with pos_tare, and other module.
Kind regards.

**Comment by robinkeunen** [comment]
Hi @Fkawala, sorry for the late response, I got myself some new eyes (operation). I'll passed it on to the colleagues.

**Comment by Fkawala** [comment]
@robinkeunen I hope your new eyes are working fine! Thank you for the message to your colleagues.

@legalsylvain Thank you, I agree with you, clean merge+refacto for v12 is the best approach. To the best I can foresee, I'll have some time to spend on v12 related work during the coming months. 

Thank you both and best regards

**Comment by vvrossem** [comment]
Hello @Fkawala, @robinkeunen told me about this PR, I will try to test and review the module this week :v:

**Comment by Fkawala** [comment]
Howdy @vvrossem thank you in advance! Feel free to reach out if I can clarify anything on this PR. Have a nice day!

File: pos_self_weighting/static/src/xml/pos_self_weighting.xml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                        <div class='subwindow-container'>
+                            <div class='subwindow-container-fix'>
+                                <div class="placeholder-OrderWidget"></div>
+                            </div>
+                        </div>
+                    </div>
+                    <div class='subwindow collapsed'>
+                        <div class='subwindow-container'>
+                            <div class='subwindow-container-fix pads'>
+                                <div class="placeholder-ActionpadWidget"></div>
```
Comment (vvrossem) [inline pos_self_weighting/static/src/xml/pos_self_weighting.xml]: Should it be `placeholder-SelfServiceActionpadWidget` instead ?

File: pos_self_weighting/static/src/xml/pos_self_weighting.xml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                    <tr class="content-row">
+                        <td class="content-cell">
+                            <div class="content-container">
+                                <span class="placeholder-ProductListWidget" />
+                            </div>
+                        </td>
+                    </tr>
+                </table>
+            </div>
+            <div class="placeholder-PriceLabelsWidget"></div>
```
Comment (vvrossem) [inline pos_self_weighting/static/src/xml/pos_self_weighting.xml]: Should it be `placeholder-SelfServiceProductListWidget` instead?

File: pos_self_weighting/static/src/js/pos_self_weighting.js [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+                    price_labels: this.price_labels});
+
+            // To replace the OrderWidget in this screen would be the regular
+            // product screen so we replace the placeholders only when this
+            // pos is configured to be a self service POS.
+            if (this.pos.config.iface_self_weight) {
+                this.order_widget.replace(this.$('.placeholder-OrderWidget'));
+                this.product_list_widget.replace(
+                    this.$('.placeholder-ProductListWidget'));
+                this.actionpad.replace(this.$('.placeholder-ActionpadWidget'));
```
Comment (vvrossem) [inline pos_self_weighting/static/src/js/pos_self_weighting.js]: Shouldn't it be `placeholder-SelfServiceActionpadWidget` instead?

File: pos_self_weighting/static/src/js/pos_self_weighting.js [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+                {order_widget: this.order_widget,
+                    price_labels: this.price_labels});
+
+            // To replace the OrderWidget in this screen would be the regular
+            // product screen so we replace the placeholders only when this
+            // pos is configured to be a self service POS.
+            if (this.pos.config.iface_self_weight) {
+                this.order_widget.replace(this.$('.placeholder-OrderWidget'));
+                this.product_list_widget.replace(
+                    this.$('.placeholder-ProductListWidget'));
```
Comment (vvrossem) [inline pos_self_weighting/static/src/js/pos_self_weighting.js]: Shouldn't it be `placeholder-SelfServiceProductListWidget ` instead?

File: pos_barcode_tare/static/src/js/pos_barcode_tare.js [hunk 5]
```diff
@@ -67,9 +65,9 @@ odoo.define('pos_barcode_tare.screens', function (require) {
             barcode_tare_action: function (code) {
                 try {
                     var order = this.pos.get_order();
-                    var last_order_line = order.get_last_orderline();
+                    var selected_order_line = order.get_selected_orderline();
```
Comment (vvrossem) [inline pos_barcode_tare/static/src/js/pos_barcode_tare.js:80]: nice !

File: pos_self_weighting/pos_self_weighting.xml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<odoo>
+    <template id="assets_backend" name="pos_barcode_tare" inherit_id="point_of_sale.assets">
+            <xpath expr="." position="inside">
+                <link rel="stylesheet" href="/pos_self_weighting/static/src/css/pos_self_weighting.css"/>
+                <script type="text/javascript" src="/pos_self_weighting/static/src/js/pos_self_weighting.js"></script>
+            </xpath>
+    </template>
+</odoo>
```
Comment (vvrossem) [inline pos_self_weighting/pos_self_weighting.xml]: This file can be renamed `assets.xml` and placed in `views` folder

File: pos_self_weighting/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    'summary': """Point of Sale - Self service weighting station for loose\
+                  goods.""",
+    'author': "Le Nid, Odoo Community Association (OCA)",
+    'website': "https://github.com/OCA/pos",
+    'license': 'AGPL-3',
+    'maintainers': ['fkawala'],
+    'depends': ['point_of_sale', 'pos_barcode_tare'],
+    'demo': ['demo/pos_self_weighting_demo.xml'],
+    'data': [
+        'pos_self_weighting.xml',
```
Comment (vvrossem) [inline pos_self_weighting/__openerp__.py]: This file can be renamed `assets.xml` and placed in `views` folder

**Review by vvrossem — COMMENTED** [review]
Hello @Fkawala,
I hope you are doing well.
Great job. It's nice to see another point of view on the self-service scale: building a POC was my internship project!

Please find below some questions/comments:
- The possibility to weigh multiple products and then print all the label at once is refreshing, but I am unsure it's compatible with printers that automatically unstick the label from the paper roll.
- For security, `close_button` in the `HeaderButtonWidget` shouldn't be accessible to customers : they could close the POS session and enter in the backend. 
- Insignificant detail, but when no electronic scale is configured, the _You did not configured this POS to use the electronic scale (...)_ popup is shown when selecting a product but not when taring a container

File: pos_barcode_tare/static/src/js/pos_barcode_tare.js (hunk 5 above)
Comment (Fkawala) [reply pos_barcode_tare/static/src/js/pos_barcode_tare.js:80]: Thanks :) it was a nice suggestion!

File: pos_self_weighting/static/src/xml/pos_self_weighting.xml (hunk 1 above)
Comment (Fkawala) [reply pos_self_weighting/static/src/xml/pos_self_weighting.xml]: Yep, good point, thanks, I'll fix that.

File: pos_self_weighting/static/src/xml/pos_self_weighting.xml (hunk 2 above)
Comment (Fkawala) [reply pos_self_weighting/static/src/xml/pos_self_weighting.xml]: Yep, good point, thanks, I'll fix that.

File: pos_self_weighting/static/src/xml/pos_self_weighting.xml (hunk 1 above)
Comment (Fkawala) [reply pos_self_weighting/static/src/xml/pos_self_weighting.xml]: Yep, good point, thanks, I'll fix that.

**Comment by vvrossem** [comment]
`self_weighing` or `self_weighting` ? :thinking:

**Comment by Fkawala** [comment]
Hello @vvrossem , I'm doing well :) I hope you're doing good as well. Thank you for the thoughtful review, this is very useful. Please find below my inline response to your comments.

> * The possibility to weigh multiple products and then print all the labels at once is refreshing, but I am unsure it's compatible with printers that automatically unstick the label from the paper roll.

Our printer put itself on hold until the user has taken the printed label. I know that all the printers out there won't behave the same. I'll add a setting field to enable/disable that feature.

> * For security, `close_button` in the `HeaderButtonWidget` shouldn't be accessible to customers: they could close the POS session and enter in the backend.

Good point, I'll disable it. For security reasons, we plan to open this POS using a dedicated odoo account with very restricted rights.

> * Insignificant detail, but when no electronic scale is configured, the _You did not configure this POS to use the electronic scale (...)_ popup is shown when selecting a product but not when taring a container.

Good catch! I'll fix that.

**Comment by Fkawala** [comment]
> `self_weighing` or `self_weighting` ? 🤔

Arg weighing and not weighting ...  will fix!

**Comment by Fkawala** [comment]
Hello @legalsylvain, I went through our discussion about merging #447 and #436 we had during the code review of #447. 

Back then you suggested we should include information needed to reconstruct the gross weight on each order line. I think this requirement is now fulfilled. The tare weight is now displayed on each order line. This is also the case for the `pos_barcode_tare` addon. 

FWIW the barcode label also shows the net weight and the tare weight (see example below).

barcode


I would like to get started on the migration of this addon to v12. As you suggested we could refactorize code from this PR, the  `pos_barcode_tare` addon and your #436 PR in one single addon for v12. I'm not familiar enough with your code to evaluate how complex the refactorization will be. Would you have any suggestions on how we/I should proceed to get the refactorization work started? 

Thank you and have a nice rest of the weekend!

**Comment by Fkawala** [comment]
Hello @vvrossem I hope you're doing good. By any chance would you have five extra minutes to spend reviewing the change I did? All the issues you've raised are solved and I did a fair bit of work on the documentation.

**Comment by Fkawala** [comment]
Howdy @ivantodorovich, would you have some time to take a quick look at this PR? No problem if you don't! Thank you a lot!

**Comment by Fkawala** [comment]
Hello @vvrossem, I'm trying to make this add-on work with a label printer (eg. zebra [gk4200](https://www.zebra.com/fr/fr/products/printers/desktop/advanced-desktop-printers.html)) that would be connected via USB to the self-service weighing station. I see two options:

1. Define a new add-on **hw_printer**. That add-on would be based on **hw_escpos** but would be a mere wrapper around pycups. That add-on would have a single `/print`HTTP verb and would contain no specific logic related to the self-weighing task.

2. Hack my way in using [base_report_to_printer](https://github.com/OCA/report-print-send/tree/12.0/base_report_to_printer) to trigger a printing on the label printer, but that entails numerous questions I have no answers for.

Could you please share your thoughts on this? Am I missing a third option there? If not which option would you vote for?

Thank you in advance for your guidance and time!

_ps; @legalsylvain @robinkeunen any comments will be warmly welcomed._

**Comment by Fkawala** [comment]
> Hello @vvrossem, I'm trying to make this add-on work with a label printer (eg. zebra [gk4200](https://www.zebra.com/fr/fr/products/printers/desktop/advanced-desktop-printers.html)) that would be connected via USB to the self-service weighing station. I see two options:
> 
> 1. Define a new add-on **hw_printer**. That add-on would be based on **hw_escpos** but would be a mere wrapper around pycups. That add-on would have a single `/print`HTTP verb and would contain no specific logic related to the self-weighing task.
> 2. Hack my way in using [base_report_to_printer](https://github.com/OCA/report-print-send/tree/12.0/base_report_to_printer) to trigger a printing on the label printer, but that entails numerous questions I have no answers for.
> 
> Could you please share your thoughts on this? Am I missing a third option there? If not which option would you vote for?
> 
> Thank you in advance for your guidance and time!
> 
> _ps; @legalsylvain @robinkeunen any comments will be warmly welcomed._

@vvrossem I'm starting to work on this today and will implement option 1. Please do let me know if you would not pursue option 1. Thank you and have a great day.

**Comment by vvrossem** [comment]
@Fkawala I would definitely pursue option 1. When I was working on a self-service scale POC, I noted the importance to build a `hw_zpl2` module to load on a posbox[less] in order to communicate with a zebra printer. Thank you so much for your time and contribution on this !
PS : I will review this PR at the end of the day, I haven't had found the time so far

**Comment by Fkawala** [comment]
> @Fkawala I would definitely pursue option 1. When I was working on a self-service scale POC, I noted the importance to build a `hw_zpl2` module to load on a posbox[less] in order to communicate with a zebra printer. Thank you so much for your time and contribution on this !
> PS : I will review this PR at the end of the day, I haven't had found the time so far

Hello @vvrossem! I'm happy to learn I'm heading in the good direction :) No rush for the review this work is on hold until I figure out the label printer part of it. Have a nice day!

**Comment by Fkawala** [comment]
Hello, @legalsylvain I see you contributed to https://github.com/akretion/pywebdriver/. I think this is exactly what I need to send raw data to the label printer. Would you have any thoughts to share on that? Would it be a problem for this addon to depend on pywebdriver to send raw data with the label printer? Thank you in advance for your inputs!

**Comment by legalsylvain** [comment]
Hi @Fkawala. Indeed, pywebdriver is a possibility. Well, in fact for each communication between Odoo and a device, there is two part : a front end part and the hardware part. The front end communicates the hardware part, with an http request. For exemple 
- for the TMT20 : the front end (``point_of_sale``) send information to the hardware (``hw_escpos``).
- for the payment terminal : the front end ([pos_payment_terminal](https://github.com/OCA/pos/tree/12.0/pos_payment_terminal)) send information to the hardware ([hw_telium_payment_terminal](https://github.com/OCA/pos/tree/12.0/hw_telium_payment_terminal))

for each connexion, pywebdriver can replace the ``hw_`` part.

So, in this module ``pos_self_weighing`` you could make http request, and then in a hardware part  handle the communication : 
-> via an odoo module ``hw_zp12`` that will be installed in you Iot / Pos Box.
-> or via a new feature in pywebdriver.

Please before develop hardware communication, please ping akretion team, that have still worked on zebra. see for exemple @PierrickBrun comment here : https://github.com/akretion/pywebdriver/pull/49#issue-379060049

kind regards.

**Comment by Fkawala** [comment]
> Hi @Fkawala. Indeed, pywebdriver is a possibility. Well, in fact for each communication between Odoo and a device, there is two part : a front end part and the hardware part. The front end communicates the hardware part, with an http request. For exemple
> 
> * for the TMT20 : the front end (`point_of_sale`) send information to the hardware (`hw_escpos`).
> * for the payment terminal : the front end ([pos_payment_terminal](https://github.com/OCA/pos/tree/12.0/pos_payment_terminal)) send information to the hardware ([hw_telium_payment_terminal](https://github.com/OCA/pos/tree/12.0/hw_telium_payment_terminal))
> 
> for each connexion, pywebdriver can replace the `hw_` part.
> 
> So, in this module `pos_self_weighing` you could make http request, and then in a hardware part handle the communication :
> -> via an odoo module `hw_zp12` that will be installed in you Iot / Pos Box.
> -> or via a new feature in pywebdriver.
> 
> Please before develop hardware communication, please ping akretion team, that have still worked on zebra. see for exemple @PierrickBrun comment here : [akretion/pywebdriver#49 (comment)](https://github.com/akretion/pywebdriver/pull/49#issue-379060049)
> 
> kind regards.

Hello Sylvain,

Thanks for the thoughtful answer, that is very much appreciated. 

My understanding (was/still is) that the `pywebdriver` has a CUPS driver. I plan to use that driver as it, without any modification. I should be able to do that because the zebra printer works with CUPS "out of the box". The only caveat is to use the ZPL-II language and configure pycups to send raw data. The ZPL-II encoded label would be generated in this module (pos_self_weighing).

Can you foresee any problem with that approach? 

Thank you & have a nice day

**Comment by PierrickBrun** [comment]
Hi @Fkawala we already did this to print labels using ZPL on a Zebra printer, so this is a working approach.

We used https://github.com/akretion/ak-odoo-incubator/tree/10.0/proxy_action to help us doing that

**Comment by Fkawala** [comment]
> Hi @Fkawala we already did this to print labels using ZPL on a Zebra printer, so this is a working approach.
> 
> We used https://github.com/akretion/ak-odoo-incubator/tree/10.0/proxy_action to help us doing that

Thank you @PierrickBrun, great to know I'm not the first one going that way! Have a nice day!

**Comment by legalsylvain** [comment]
Hi. @pierrickBrun do you mean this module will depend on proxy_action ?
If yes, it should be in OCA too. Is it planned ?

Regarding the design. This module will work in a first time only with pywebdriver that is not under the OCA umbrella. (But quasi : developped by GRAP, Akretion, Acsone, that are active members).
If such design is selected and if i understand correctly, you should write in the roadmap that a hw_pz should be developped.
Kind regards.

**Comment by Fkawala** [comment]
@legalsylvain Do you think we should rather have code duplication and stay under the OCA umbrella? To me, it does **not** look like a lot of work. I'm keen on having a minimal piece of code (a `hw_xyz` module, implying code duplication), and a roadmap item to switch to pywebdriver whenever it has been moved under the OCA umbrella. I'm very open to whatever solution, but I like to wrap up the work in a relatively short timeline. We would like to try the first version in our shop by the end of the month.

**Comment by legalsylvain** [comment]
Not sure to understand correctly. 
I was proposing only to developp the ``pos_self_weighting`` module, calling pywebdriver, and adding in the roadmap that it could be great to developp the ``hw_`` module for people that don't use ``pywebdriver``.
But a lot of module are depending on extra python libraries or package. So it's OK I think.
regarding ``proxy_action``, if your module has to depend on it, yes it should be under OCA umbrella, but I don't think it's a problem. @PierrickBrun : what is the roadmap, regarding this incubated module ?

kind regards.

**Comment by legalsylvain** [comment]
Hi @Fkawala. don't know if you know this module, but you maybe should take a look on this module : 
https://github.com/coopiteasy/pos/pull/8

**Comment by legalsylvain** [comment]
this PR is still active ?

**Comment by Fkawala** [comment]
> this PR is still active ?

Nop, I'll do a new one for v12.0 only. The new code will depend on `pos_tare` and `pos_tare_generate_barcode`. I'll close this. Thank you for asking :)

**Pull request closed without merging**